File: daily/Arrays/948.Bag-of-tokens.go
package daily

import "sort"

func bagOfTokensScore(tokens []int, power int) int {
	score := 0
	sort.Ints(tokens)
	left, right := 0, len(tokens)-1

	if len(tokens) == 0 {
		return 0
	}

	for left < right {
		if score == 0 {
			if power >= tokens[left] {
				score += 1
				power = power - tokens[left]
			}
			left++
		} else {
			if power < tokens[left] {
				power = power + tokens[right]
				score -= 1
				right--
			} else {
				score += 1
				power = power - tokens[left]
				left++
			}
		}
	}

	if tokens[right] <= power {
		score++
	}

	return score
}

File: daily/Arrays/948.Bag-of-tokens.py
from typing import List


class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        tokens.sort()
        score = 0
        best = 0
        left, right = 0, len(tokens) - 1

        while left <= right and (power >= tokens[left] or score > 0):
            if power >= tokens[left]:
                power -= tokens[left]
                score += 1
                best = max(best, score)
                left += 1
            elif score > 0:
                power += tokens[right]
                score -= 1
                right -= 1

        return best
